worker/src/RTC/ActiveSpeakerObserver.cpp

```diff
@@ -0,0 +1,523 @@
+		                       (nR - vL) * std::log(1 - p) - std::log(lambda) + lambda * vL;
+
+		if (activityScore < MinActivityScore)
+		{
+			activityScore = MinActivityScore;
+		}
+
+		return activityScore;
+	}
+
+	inline bool ComputeBigs(
+	  const std::vector<uint8_t>& littles, std::vector<uint8_t>& bigs, uint8_t threashold)
+	{
+		uint32_t littleLen       = littles.size();
+		uint32_t bigLen          = bigs.size();
+		uint32_t littleLenPerBig = littleLen / bigLen;
+		bool changed             = false;
+
+		for (uint32_t b = 0, l = 0; b < bigLen; b++)
+		{
+			uint8_t sum = 0;
+
+			for (uint32_t lEnd = l + littleLenPerBig; l < lEnd; l++)
+			{
+				if (littles[l] > threashold)
+				{
+					sum++;
+				}
+			}
+			if (bigs[b] != sum)
+			{
+				bigs[b] = sum;
+				changed = true;
+			}
+		}
+		return changed;
+	}
+
+	ActiveSpeakerObserver::ActiveSpeakerObserver(const std::string& id, json& data)
+	  : RTC::RtpObserver(id)
+	{
+		MS_TRACE();
+
+		auto jsonIntervalIt = data.find("interval");
+
+		if (jsonIntervalIt == data.end() || !jsonIntervalIt->is_number())
+			MS_THROW_TYPE_ERROR("missing interval");
+
+		this->interval = jsonIntervalIt->get<int16_t>();
+
+		if (this->interval < 100)
+			this->interval = 100;
+		else if (this->interval > 5000)
+			this->interval = 5000;
+
+		this->periodicTimer = new Timer(this);
+
+		this->periodicTimer->Start(interval, interval);
+	}
+
+	ActiveSpeakerObserver::~ActiveSpeakerObserver()
+	{
+		MS_TRACE();
+
+		delete this->periodicTimer;
+	}
+
+	void ActiveSpeakerObserver::AddProducer(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		if (producer->GetKind() != RTC::Media::Kind::AUDIO)
+			MS_THROW_TYPE_ERROR("not an audio Producer");
+
+		this->mapProducerSpeaker[producer->id].producer = producer;
+		this->mapProducerSpeaker[producer->id].speaker  = new Speaker();
+	}
+
+	void ActiveSpeakerObserver::RemoveProducer(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		if (mapProducerSpeaker[producer->id].speaker != nullptr)
+		{
+			delete mapProducerSpeaker[producer->id].speaker;
+		}
+
+		this->mapProducerSpeaker.erase(producer->id);
+
+		if (producer->id == this->dominantId)
+		{
+			Update();
+		}
+	}
+
+	void ActiveSpeakerObserver::ProducerResumed(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		if (producer->GetKind() != RTC::Media::Kind::AUDIO)
+			MS_THROW_TYPE_ERROR("not an audio Producer");
+
+		this->mapProducerSpeaker[producer->id].speaker->paused = false;
+	}
+
+	void ActiveSpeakerObserver::ProducerPaused(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		this->mapProducerSpeaker[producer->id].speaker->paused = true;
+	}
+
+	void ActiveSpeakerObserver::ReceiveRtpPacket(RTC::Producer* producer, RTC::RtpPacket* packet)
+	{
+		MS_TRACE();
+
+		if (IsPaused())
+			return;
+
+		uint8_t volume;
+		bool voice;
+
+		if (!packet->ReadSsrcAudioLevel(volume, voice))
+			return;
+
+		uint64_t now = DepLibUV::GetTimeMs();
+		this->mapProducerSpeaker[producer->id].speaker->LevelChanged(volume, now);
```

Comment:
What if a packet is received from a Producer that is not present in this map? This will crash.